src/sqlfluff/core/linter.py

```diff
@@ -105,275 +101,136 @@ def is_clean(self):
     def fix_string(self):
         """Obtain the changes to a path as a string.
 
-        We use the file_mask to do a safe merge, avoiding any templated
-        sections. First we need to detect where there have been changes
-        between the fixed and templated versions. The file mask is of
-        the format: (raw_file, templated_file, fixed_file).
-
-        We use difflib.SequenceMatcher.get_opcodes
-        See: https://docs.python.org/3.7/library/difflib.html#difflib.SequenceMatcher.get_opcodes
-        It returns a list of tuples ('equal|replace|delete|insert', ia1, ia2, ib1, ib2).
+        We use the source mapping features of TemplatedFile
+        to generate a list of "patches" which cover the non
+        templated parts of the file and refer back to the locations
+        in the original file.
 
+        NB: This is MUCH FASTER than the original approach
+        using difflib in pre 0.4.0.
         """
         bencher = BenchIt()
         bencher("fix_string: start")
 
-        # Do we have enough information to actually fix the file?
-        if any(elem is None for elem in self.file_mask):
-            linter_logger.warning(
-                "Insufficient information to fix file: %s", self.file_mask
-            )
-            return None, False
+        linter_logger.debug("Original Tree: %r", self.templated_file.templated_str)
+        linter_logger.debug("Fixed Tree: %r", self.tree.raw)
 
-        linter_logger.info("Persisting file masks: %s", self.file_mask)
-        # Compare Templated with Raw
-        diff_templ = SequenceMatcher(
-            autojunk=None, a=self.file_mask[0], b=self.file_mask[1]
-        )
-        bencher("fix_string: Match 0&1")
-        diff_templ_codes = diff_templ.get_opcodes()
-        linter_logger.debug("Templater diff codes: %s", diff_templ_codes)
-
-        bencher("fix_string: Got Opcodes 0&1")
-        # Compare Fixed with Templated
-        diff_fix = SequenceMatcher(
-            autojunk=None, a=self.file_mask[1], b=self.file_mask[2]
+        # The sliced file is contigious in the TEMPLATED space.
+        # NB: It has gaps and repeats in the source space.
+        # It's also not the FIXED file either.
+        linter_logger.debug("### Templated File.")
+        for idx, file_slice in enumerate(self.templated_file.sliced_file):
+            linter_logger.debug("    File slice: %s %r", idx, file_slice)
+
+        original_source = self.templated_file.source_str
+
+        # Make sure no patches overlap and divide up the source file into slices.
+        # Any Template tags in the source file should be "untouchable".
+        untouchable_slices = self.templated_file.untouchable_slices()
```

Comment:
Not sure about the `untouchable_slice` concept, if it is only for templated parts, couldn't we name it `template_slice`?

Reply:
Ha - yeah that's probably a much better name! I'll see what I can do on that front.

Reply:
On reflection, it's not just _templated slices_, but a subset of them which should be expected to have zero-length in the templated file (i.e. comments or blocks). I've renamed the method to `source_only_slices` and updated the signature to return `RawFileSlice` objects directly.